spopt/region/skater.py

```diff
@@ -0,0 +1,432 @@
+        """
+        if MSF is None:
+            MSF = self.minimum_spanning_forest_
+        if np.isfinite(score):
+            MSF[in_node, out_node] = 0
+            MSF.eliminate_zeros()
+            return (MSF, *cg.connected_components(MSF, directed=False))
+        raise OptimizeWarning(
+            "Score of the ({},{}) cut is inf, the quorum is likely not met!"
+        )
+
+
+class Skater(BaseSpOptHeuristicSolver):
+    """Skater is a spatial regionalization algorithm based on spanning tree pruning"""
+
+    def __init__(
+        self,
+        gdf,
+        w,
+        attrs_name,
+        n_clusters=5,
+        floor=-np.inf,
+        trace=False,
+        islands="increase",
+        dissimilarity=skm.manhattan_distances,
+        affinity=None,
+        reduction=np.sum,
+        center=np.mean,
+    ):
+        """
+        # SKATER spatial clustering algorithm.
+        Parameters
+        ----------
+        
+        gdf : geopandas.GeoDataFrame
+            ... : 
+   
+        w : libpysal.weights.W instance
+        spatial weights matrix
+
+        attrs_name : list
+        Strings for attribute names (cols of ``geopandas.GeoDataFrame``).
+        
+        n_clusters : int, optional, default: 5
+        The number of clusters to form.
+
+        floor: floor on the size of regions.
+
+        trace: bool denoting whether to store intermediate
+               labelings as the tree gets pruned
+               
+        islands: string describing what to do with islands. 
+                 If "ignore", will discover `n_clusters` regions, treating islands as their own regions.
+                 If "increase", will discover `n_clusters` regions, treating islands as separate from n_clusters. 
+
+
+        Parameters for spanning tree pruning.
+        ----------
+
+        dissimilarity : a callable distance metric.
+
+        affinity : an callable affinity metric between 0,1. 
+                   Will be inverted to provide a 
+                   dissimilarity metric.
+
+        reduction: the reduction applied over all clusters
+                   to provide the map score.
+
+        center:    way to compute the center of each region in attribute space
+        
+        """
+        self.gdf = gdf
+        self.w = w
+        self.attrs_name = attrs_name
+        self.n_clusters = n_clusters
+        self.floor = floor
+        self.trace = trace
+        self.islands = islands
+        self.dissimilarity = dissimilarity
+        self.affinity = affinity
+        self.reduction = reduction
+        self.center = center
+
+    def solve(self):
+        """Solve the region k-means heuristic."""
+        data = self.gdf
+        X = data[self.attrs_name].values
+        model = SpanningForest(self.dissimilarity, self.affinity, self.reduction, self.center)
+        model.fit(self.n_clusters, self.w, data=X, quorum=self.floor, trace=self.trace)
+        self.labels_ = model.current_labels_
+
+
+if __name__ == "__main__":
```

Comment:
I am not sure there is a reason for having this last section since the tests are happening in `tests/test_skater.py`. Is there a reason for keeping it that I am not seeing?

Reply:
Thanks, I just removed it.